*: fix numRows telemetry for bulk jobs (IMPORT, BACKUP, RESTORE)

ZhouXing19 opened the pull request:
Previously, the `numRows` entry for the sampled query event of a bulk job is always 1. 
It's not reflecting the true number of changed rows. This is because, for a
sampled query event, the `NumRows` entry is from the `commandResult.rowsAffected`
entry, which only increases by one when there's only 1 returned row from the
command. This becomes tricker when the bulk job query is with the `detached` option,
where the telemetry logging happens **BEFORE** the job getting executed.

Hence for these 3 special kinds of commands, we emit the number of changed rows
in the log when the job is completed (succeeded, canceled or failed) in a job recovery
event. To link the sampled query event with it , we also print the BulkJobId along
with the sampled query.

For example, an IMPORT command will give
```log
{..."EventType":"sampled_query","Statement":"IMPORT INTO defaultdb.public.test(x) CSV DATA (‹'nodelocal://1/test.csv'›) WITH detached","Tag":"IMPORT",...,"BulkJobId":831781983466061825,...}
{..."EventType":"recovery_event","RecoveryType":"import","JobID":831781983466061825,"ResultStatus":"succeeded","NumRows":8}
```

Note that we moved `pkg/sql/telemetry_logging_test.go` to `sql_test` package,
just to avoid the dependency loop between `sql` and `nodelocal`.

fixes #91940

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/95042)
<!-- Reviewable:end -->

ZhouXing19 committed:
*: fix numRows telemetry for bulk jobs (IMPORT, BACKUP, RESTORE)

fixes #91940

Previously, the numRows entry for the sampled query event of a bulk job is always
1. It's not reflecting the true number of changed rows. This is because for a
sampled query event, the `NumRows` entry is from the `commandResult.rowsAffected`
entry, which only increase by one when there's only 1 returned row from the
command. This become tricker when the bulk job query is with the `detached` option,
where the telemetry logging happens BEFORE the job got executed.

Hence for these 3 special kinds of commands, we emit the number of changed rows
in the log when the job is completed (succeeded, canceled or failed) in a job recovery
event. To link the sampled query event with it , we also print the BulkJobId along
with the sampled query.

For example, an IMPORT command will give
```log
{..."EventType":"sampled_query","Statement":"IMPORT INTO defaultdb.public.test(x) CSV DATA (‹'nodelocal://1/test.csv'›) WITH detached","Tag":"IMPORT",...,"BulkJobId":831781983466061825,...}
{..."EventType":"recovery_event","RecoveryType":"import","JobID":831781983466061825,"ResultStatus":"succeeded","NumRows":8}
```

Release note: None

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95042)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ZhouXing19)

___
*[`pkg/sql/telemetry_logging_test.go` line 1237 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NN7jhLMCnhjabubSNk5:-NN7jhLMCnhjabubSNk6:buzbcaz) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e7483635aa015210a4170eacb94927a34cffcd7/pkg/sql/telemetry_logging_test.go#L1237)):*
> ```Go
> }
> 
> // TODO(sql-exp or dr): add event telemetry tests for failed or canceled bulk jobs.
> ```

nit: the name inside a TODO is supposed to be someone that you can ask about the respective note. It's unusual to put a team there.
___
*[`pkg/sql/telemetry_logging_test.go` line 1407 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NN7ilNT0HzVqBgKuKTB:-NN7ilNT0HzVqBgKuKTC:b-lgqykj) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e7483635aa015210a4170eacb94927a34cffcd7/pkg/sql/telemetry_logging_test.go#L1407)):*
> ```Go
> 
> 	var filteredSampleQueries []logpb.Entry
> 	if err := retry.WithMaxAttempts(context.Background(), retry.Options{}, 500, func() error {
> ```

you know about `testutils.SucceedsSoon()`, right?
___
*[`pkg/sql/telemetry_logging_test.go` line 1459 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NN7jBHV80Hqph-1VkU7:-NN7jBHV80Hqph-1VkU8:bks085r) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e7483635aa015210a4170eacb94927a34cffcd7/pkg/sql/telemetry_logging_test.go#L1459)):*
> ```Go
> 	}
> 
> 	if len(recoveryEventEntries) == 0 {
> ```

`require.Equal` ?
___
*[`pkg/sql/telemetry_logging_test.go` line 1463 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NN7jK4W40Lfjt-cBb3c:-NN7jK4W40Lfjt-cBb3d:b-l9d628) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e7483635aa015210a4170eacb94927a34cffcd7/pkg/sql/telemetry_logging_test.go#L1463)):*
> ```Go
> 	}
> 
> 	if err != nil {
> ```

what error is this checking? Is this the right place for it
___
*[`pkg/sql/telemetry_logging_test.go` line 1464 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NN7jCxyAb-sX5VQnec6:-NN7jCxyAb-sX5VQnec7:b-82dbbs) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e7483635aa015210a4170eacb94927a34cffcd7/pkg/sql/telemetry_logging_test.go#L1464)):*
> ```Go
> 
> 	if err != nil {
> 		t.Fatal(err)
> ```

`require.NoError` (and elsewhere) ?
___
*[`pkg/sql/telemetry_logging_test.go` line 1478 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NN7jWs940peCQda9dQN:-NN7jWs940peCQda9dQO:b-o8shoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e7483635aa015210a4170eacb94927a34cffcd7/pkg/sql/telemetry_logging_test.go#L1478)):*
> ```Go
> 				}
> 				stmt := sq.Statement.StripMarkers()
> 				_ = stmt
> ```

what's going on here?


<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
The CI still fails with 
```
    telemetry_logging_test.go:1503: Leaked goroutine: goroutine 26 [select]:
        go.opencensus.io/stats/view.(*worker).start(0xc000da5e00)
                go.opencensus.io/stats/view/external/io_opencensus_go/stats/view/worker.go:276 +0xad
        created by go.opencensus.io/stats/view.init.0
                go.opencensus.io/stats/view/external/io_opencensus_go/stats/view/worker.go:34 +0x8d
```
on the newly added test, and I'm still investigating (see also this [internal link](https://cockroachlabs.slack.com/archives/C9TGBJB44/p1675192660786999)). But I think the code is ready for a look now.

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95042)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @DrewKimball, and @msbutler)

___
*[`pkg/sql/telemetry_logging_test.go` line 1237 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NN7jhLMCnhjabubSNk5:-NN86RwaEZCdJ5zlljJa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e7483635aa015210a4170eacb94927a34cffcd7/pkg/sql/telemetry_logging_test.go#L1237)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: the name inside a TODO is supposed to be someone that you can ask about the respective note. It's unusual to put a team there.
</blockquote></details>

Done.
___
*[`pkg/sql/telemetry_logging_test.go` line 1407 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NN7ilNT0HzVqBgKuKTB:-NN86TP33Wcd9F8_89Mv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e7483635aa015210a4170eacb94927a34cffcd7/pkg/sql/telemetry_logging_test.go#L1407)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

you know about `testutils.SucceedsSoon()`, right?
</blockquote></details>

Done.
___
*[`pkg/sql/telemetry_logging_test.go` line 1459 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NN7jBHV80Hqph-1VkU7:-NN86VT39g21KTXSaGqf:b-47svpw) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e7483635aa015210a4170eacb94927a34cffcd7/pkg/sql/telemetry_logging_test.go#L1459)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

`require.Equal` ?
</blockquote></details>

I put it inside `testutils.SucceedsSoon()`, so I think we change it to return an error.
___
*[`pkg/sql/telemetry_logging_test.go` line 1463 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NN7jK4W40Lfjt-cBb3c:-NN86i1xCYaIJgNc3fEz:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e7483635aa015210a4170eacb94927a34cffcd7/pkg/sql/telemetry_logging_test.go#L1463)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

what error is this checking? Is this the right place for it
</blockquote></details>

Removed.
___
*[`pkg/sql/telemetry_logging_test.go` line 1464 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NN7jCxyAb-sX5VQnec6:-NN86kEI98rsU8W2X4aN:bi681ti) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e7483635aa015210a4170eacb94927a34cffcd7/pkg/sql/telemetry_logging_test.go#L1464)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

`require.NoError` (and elsewhere) ?
</blockquote></details>

Ditto, changed it to returning an error.
___
*[`pkg/sql/telemetry_logging_test.go` line 1478 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NN7jWs940peCQda9dQN:-NN86mkr3zSLAXvh3MBI:b5c4q3a) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e7483635aa015210a4170eacb94927a34cffcd7/pkg/sql/telemetry_logging_test.go#L1478)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

what's going on here?
</blockquote></details>

Ah this was for the debugger -- this PR hasn't been cleaned up yet.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95042)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @msbutler, @rafiss, and @ZhouXing19)

___
*[`pkg/sql/telemetry_logging_test.go` line 1443 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NN8sZYz8kwWRSqiCpVe:-NN8sZYz8kwWRSqiCpVf:b-8mblt8) ([raw file](https://github.com/cockroachdb/cockroach/blob/7774e7389c782761d7e649cb36928a0217d0cba1/pkg/sql/telemetry_logging_test.go#L1443)):*
> ```Go
> 		)
> 		if err != nil {
> 			return err
> ```

do you expect this to ever fail? If not, `require.NoError` (even inside a `SucceedsSoon`).
___
*[`pkg/sql/telemetry_logging_test.go` line 1446 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NN8s75b5QyiCdV7OgeG:-NN8s75b5QyiCdV7OgeH:b-a6eofy) ([raw file](https://github.com/cockroachdb/cockroach/blob/7774e7389c782761d7e649cb36928a0217d0cba1/pkg/sql/telemetry_logging_test.go#L1446)):*
> ```Go
> 		}
> 		if len(recoveryEventEntries) < len(testData) {
> 			time.Sleep(500 * time.Millisecond)
> ```

do you still need this sleep? `SucceedsSoon` does an exponential backoff.
___
*[`pkg/sql/telemetry_logging_test.go` line 1447 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NN8sSYz6SCeY26aUizs:-NN8sSYz6SCeY26aUizt:b1rhwti) ([raw file](https://github.com/cockroachdb/cockroach/blob/7774e7389c782761d7e649cb36928a0217d0cba1/pkg/sql/telemetry_logging_test.go#L1447)):*
> ```Go
> 		if len(recoveryEventEntries) < len(testData) {
> 			time.Sleep(500 * time.Millisecond)
> 			t.Logf("found %d recovery entry, need %d\n", len(recoveryEventEntries), len(testData))
> ```

`SucceedsSoon` does some logging, fwiw
___
*[`pkg/sql/telemetry_logging_test.go` line 1451 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NN8skbF4t7cZvW8Z1UJ:-NN8skbF4t7cZvW8Z1UK:bc8hq4s) ([raw file](https://github.com/cockroachdb/cockroach/blob/7774e7389c782761d7e649cb36928a0217d0cba1/pkg/sql/telemetry_logging_test.go#L1451)):*
> ```Go
> 		}
> 		if len(recoveryEventEntries) == 0 {
> 			return errors.New("no recovery event entries found")
> ```

is hitting this possible given the `if len(recoveryEventEntries) < len(testData) {` above?


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95042)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @DrewKimball, @msbutler, and @rafiss)

___
*[`pkg/sql/telemetry_logging_test.go` line 1443 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NN8sZYz8kwWRSqiCpVe:-NN8twW81kWYaSPPy7tQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7774e7389c782761d7e649cb36928a0217d0cba1/pkg/sql/telemetry_logging_test.go#L1443)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

do you expect this to ever fail? If not, `require.NoError` (even inside a `SucceedsSoon`).
</blockquote></details>

Done.
___
*[`pkg/sql/telemetry_logging_test.go` line 1446 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NN8s75b5QyiCdV7OgeG:-NN8txCS-VcujRcg--M9:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/7774e7389c782761d7e649cb36928a0217d0cba1/pkg/sql/telemetry_logging_test.go#L1446)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

do you still need this sleep? `SucceedsSoon` does an exponential backoff.
</blockquote></details>

Removed.
___
*[`pkg/sql/telemetry_logging_test.go` line 1447 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NN8sSYz6SCeY26aUizs:-NN8ty8i-f9vFp9PMIQY:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/7774e7389c782761d7e649cb36928a0217d0cba1/pkg/sql/telemetry_logging_test.go#L1447)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

`SucceedsSoon` does some logging, fwiw
</blockquote></details>

Removed.
___
*[`pkg/sql/telemetry_logging_test.go` line 1451 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NN8skbF4t7cZvW8Z1UJ:-NN8tyyV2_J87Xl9HzMx:b-89b8ez) ([raw file](https://github.com/cockroachdb/cockroach/blob/7774e7389c782761d7e649cb36928a0217d0cba1/pkg/sql/telemetry_logging_test.go#L1451)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

is hitting this possible given the `if len(recoveryEventEntries) < len(testData) {` above?
</blockquote></details>

Right, removed this.


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
you can exclude that stack from the leak detector here: https://github.com/cockroachdb/cockroach/blob/d6a32ed664a3cb6eb542277b867ddc4ce25d53d4/pkg/util/leaktest/leaktest.go#L39

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-:-NND8hIk6QMmSeJrWJAK:bigf041)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @DrewKimball, @msbutler, and @ZhouXing19)

___
*[`docs/generated/eventlog.md` line 2812 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NND9Dqi-Z94ormVMiNl:-NND9Dqi-Z94ormVMiNm:b-p3j9k1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d1147f337155975914027ccf1514643242fd6ca/docs/generated/eventlog.md#L2812)):*
> ```Markdown
> | `IgnoreExistingBackup` | IgnoreExistingBackup is true iff the BACKUP schedule should still be created even if a backup is already present in its destination. | no |
> | `ApplicationName` | The application name for the session where recovery event was created. | no |
> | `NumRows` | NumRows is the number of rows successfully imported, backup-ed or restored. | no |
> ```

nit: "backup-ed" should be "backed up"
___
*[`pkg/sql/conn_io.go` line 992 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NND9wVV09mwtCQBioNH:-NND9wVV09mwtCQBioNI:bi7wwg9) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d1147f337155975914027ccf1514643242fd6ca/pkg/sql/conn_io.go#L992)):*
> ```Go
> // GetEntryFromExtraInfo is part of the sql.RestrictedCommandResult interface.
> func (r *streamingCommandResult) GetBulkJobId() uint64 {
> 	panic("unimplemented")
> ```

could this just return 0?
___
*[`pkg/sql/exec_log.go` line 326 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NNDA4dg-QIQkzYePyvP:-NNDA4dg-QIQkzYePyvQ:brc3wrv) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d1147f337155975914027ccf1514643242fd6ca/pkg/sql/exec_log.go#L326)):*
> ```Go
> 	// print out the number of changed rows along with the sampled query event.
> 	// We emit it when the job succeeds in a recovery_event.
> 	var dontPrintNumRows bool
> ```

it's a bit confusing to have a boolean variable with a negation in the name. could you instead intialize this as:
```
printNumRows := true
```

and change it to false if needed

or better yet, simplify the whole thing:
```
	switch p.stmt.AST.(type) {
	case *tree.Import, *tree.Restore, *tree.Backup:
		execDetails.BulkJobId = bulkJobId
	default:
		execDetails.NumRows = int64(rows)
	}
```

so there's no need for the `bool`
___
*[`pkg/sql/importer/import_job.go` line 1398 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NNDAm8d8FFgDLjbsN_-:-NNDAm8d8FFgDLjbsN_0:b256mb1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d1147f337155975914027ccf1514643242fd6ca/pkg/sql/importer/import_job.go#L1398)):*
> ```Go
> 	emitImportJobEvent(ctx, p, jobs.StatusReverting, r.job)
> 
> 	// TODO(sql-exp): increase telemetry count for import.total.failed and
> ```

could you explain what "increase telemetry count" means?
___
*[`pkg/sql/pgwire/command_result.go` line 228 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NNDCdX42uWk7PinkIF5:-NNDCdX42uWk7PinkIF6:b5rc3n7) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d1147f337155975914027ccf1514643242fd6ca/pkg/sql/pgwire/command_result.go#L228)):*
> ```Go
> const (
> 	jobIdColIdx = 0
> 	rowsColIdx  = 3
> ```

it seems like `rowsColIdx` is unused except for comments. do we need it?
___
*[`pkg/sql/pgwire/command_result.go` line 238 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NNDEFlH6TaOBwVz-ieO:-NNDEFlH6TaOBwVz-ieP:b8tg6ck) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d1147f337155975914027ccf1514643242fd6ca/pkg/sql/pgwire/command_result.go#L238)):*
> ```Go
> 	switch r.cmdCompleteTag {
> 	case tree.ImportTagStr, tree.RestoreTagStr, tree.BackupTagStr:
> 		if row.Len() == 0 {
> ```

is this length check really necessary?
___
*[`pkg/sql/pgwire/command_result.go` line 244 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NNDENY9-mREheav2h9e:-NNDENY9-mREheav2h9f:b-b3rxn9) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d1147f337155975914027ccf1514643242fd6ca/pkg/sql/pgwire/command_result.go#L244)):*
> ```Go
> 		// jobs.DetachedJobExecutionResultHeader, the returned row can have either
> 		// 1 or 6 entries. The first entry is always the job id.
> 		// If there are 6 entries, it must contain "rows".
> ```

the thing about 1 or 6 entries doesn't seem that relevant for this comment, all that matters is the first entry

also it would be nice if there was a different way to get the job ID, since this seems a bit brittle if names or indexes change
___
*[`pkg/sql/pgwire/command_result.go` line 259 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NNDEWwB1DV6raQkmdHt:-NNDEWwB1DV6raQkmdHu:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d1147f337155975914027ccf1514643242fd6ca/pkg/sql/pgwire/command_result.go#L259)):*
> ```Go
> 	switch r.cmdCompleteTag {
> 	case tree.ImportTagStr, tree.RestoreTagStr, tree.BackupTagStr:
> 		if batch.Width() == 0 {
> ```

ditto
___
*[`pkg/sql/sem/tree/stmt.go` line 89 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NNDEtND64uKiUWXAD6S:-NNDEtND64uKiUWXAD6T:b-3yndk0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d1147f337155975914027ccf1514643242fd6ca/pkg/sql/sem/tree/stmt.go#L89)):*
> ```Go
> 
> const (
> 	RestoreTagStr = "RESTORE"
> ```

i don't think `Str` is necessary to have in all these names
___
*[`pkg/sql/telemetry_logging_test.go` line 11 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NNDBjw0D1uGQ9qNcosm:-NNDBjw0D1uGQ9qNcosn:bau4j7q) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d1147f337155975914027ccf1514643242fd6ca/pkg/sql/telemetry_logging_test.go#L11)):*
> ```Go
> // licenses/APL.txt.
> 
> package sql_test
> ```

do you know which dependencies required this to be in a different package? is that the only reason why the knobs and `executor type` had to be public?


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95042)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @DrewKimball, @msbutler, and @rafiss)

___
*[`docs/generated/eventlog.md` line 2812 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NND9Dqi-Z94ormVMiNl:-NNHfTNl1U1tdlT53wdL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d1147f337155975914027ccf1514643242fd6ca/docs/generated/eventlog.md#L2812)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: "backup-ed" should be "backed up"
</blockquote></details>

Done.
___
*[`pkg/sql/conn_io.go` line 992 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NND9wVV09mwtCQBioNH:-NNHfUJ6C4n85JdqUbbk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d1147f337155975914027ccf1514643242fd6ca/pkg/sql/conn_io.go#L992)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

could this just return 0?
</blockquote></details>

Done.
___
*[`pkg/sql/exec_log.go` line 326 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NNDA4dg-QIQkzYePyvP:-NNHfVEy4GmwLwsC6M_4:b-q5rx7c) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d1147f337155975914027ccf1514643242fd6ca/pkg/sql/exec_log.go#L326)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

it's a bit confusing to have a boolean variable with a negation in the name. could you instead intialize this as:
```
printNumRows := true
```

and change it to false if needed

or better yet, simplify the whole thing:
```
	switch p.stmt.AST.(type) {
	case *tree.Import, *tree.Restore, *tree.Backup:
		execDetails.BulkJobId = bulkJobId
	default:
		execDetails.NumRows = int64(rows)
	}
```

so there's no need for the `bool`
</blockquote></details>

Ah right, thanks! Done.
___
*[`pkg/sql/importer/import_job.go` line 1398 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NNDAm8d8FFgDLjbsN_-:-NNHfYSe-d2QUHmsE9fh:bhtmuca) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d1147f337155975914027ccf1514643242fd6ca/pkg/sql/importer/import_job.go#L1398)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

could you explain what "increase telemetry count" means?
</blockquote></details>

It means to be consistent with `restore` and `backup`. i.e. we already have `telemetry.Count("restore.total.failed")`, `telemetry.CountBucketed("restore.duration-sec.failed",...)`,
`telemetry.Count("backup.total.failed")` and `telemetry.CountBucketed("backup.duration-sec.failed",...)`
___
*[`pkg/sql/pgwire/command_result.go` line 228 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NNDCdX42uWk7PinkIF5:-NNHfwrgDfgzzS6im4Og:b-v6s1o3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d1147f337155975914027ccf1514643242fd6ca/pkg/sql/pgwire/command_result.go#L228)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

it seems like `rowsColIdx` is unused except for comments. do we need it?
</blockquote></details>

Removed. 
___
*[`pkg/sql/pgwire/command_result.go` line 238 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NNDEFlH6TaOBwVz-ieO:-NNHg7OqE7wragVus7Wf:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d1147f337155975914027ccf1514643242fd6ca/pkg/sql/pgwire/command_result.go#L238)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

is this length check really necessary?
</blockquote></details>

Removed.



_Code quote:_
```Go
if row.Len() == 0 {
```
___
*[`pkg/sql/pgwire/command_result.go` line 244 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NNDENY9-mREheav2h9e:-NNHgDsCADmww570hW3a:b-bl3e6c) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d1147f337155975914027ccf1514643242fd6ca/pkg/sql/pgwire/command_result.go#L244)):*
Removed the comment.

> also it would be nice if there was a different way to get the job ID, since this seems a bit brittle if names or indexes change

Changed it to be set in the `init()` func. 
___
*[`pkg/sql/pgwire/command_result.go` line 259 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NNDEWwB1DV6raQkmdHt:-NNHjt9y7JCGtMbrReEH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d1147f337155975914027ccf1514643242fd6ca/pkg/sql/pgwire/command_result.go#L259)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.
___
*[`pkg/sql/sem/tree/stmt.go` line 89 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NNDEtND64uKiUWXAD6S:-NNHjs8L3meRkCxxxBPl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d1147f337155975914027ccf1514643242fd6ca/pkg/sql/sem/tree/stmt.go#L89)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i don't think `Str` is necessary to have in all these names
</blockquote></details>

Done.
___
*[`pkg/sql/telemetry_logging_test.go` line 11 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NNDBjw0D1uGQ9qNcosm:-NNHjmvg67cGhY9cGOik:b-hi644g) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d1147f337155975914027ccf1514643242fd6ca/pkg/sql/telemetry_logging_test.go#L11)):*
> do you know which dependencies required this to be in a different package?

It's `nodelocal` and `sql`. There's already a dependency between them:
![Screenshot 2023-02-02 at 10.31.44.png](https://files.reviewableusercontent.io/images/reviewable/53610260/GQ6mJylg8gfqkIY03aD5t87i/Screenshot%202023-02-02%20at%2010.31.44.png)
I need `nodelocal.MakeLocalStorageURI` for the `backup` and `restore` tests.

> is that the only reason why the knobs and executor type had to be public?

Yes


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
Per the convo [here](https://cockroachlabs.slack.com/archives/CHVV403F0/p1675181739033489) (internal link), I changed the imported package for `telemetry_logging_test.go` from `backupccl` to `ccl`. This is needed because otherwise we will encounter 
```
telemetry_logging_test.go:1394: unexpected error executing query `RESTORE DATABASE mydb FROM LATEST IN 'nodelocal://0/test1' WITH detached`: pq: a CCL binary is required to use this statement type: *tree.Restore
```
I tried stress-testing it with `ccl` locally and it haven't failed yet for 30mins. So :crossed_fingers: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-:-NNHm-2k8YayD0b0xCto:b-gospik)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @DrewKimball, @msbutler, and @rafiss)



<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95042)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @DrewKimball, @msbutler, and @rafiss)

___
*[`pkg/sql/telemetry_logging_test.go` line 11 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NNDBjw0D1uGQ9qNcosm:-NNHma7h1RpL0ntdIhi4:bwzekfc) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d1147f337155975914027ccf1514643242fd6ca/pkg/sql/telemetry_logging_test.go#L11)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

> do you know which dependencies required this to be in a different package?

It's `nodelocal` and `sql`. There's already a dependency between them:
![Screenshot 2023-02-02 at 10.31.44.png](https://files.reviewableusercontent.io/images/reviewable/53610260/GQ6mJylg8gfqkIY03aD5t87i/Screenshot%202023-02-02%20at%2010.31.44.png)
I need `nodelocal.MakeLocalStorageURI` for the `backup` and `restore` tests.

> is that the only reason why the knobs and executor type had to be public?

Yes
</blockquote></details>

This dependency is absolutely wild. We should nix it. The dependency is on the `execCfg` which is used for 3 fields: `Settings`, `ExternalIODir` and `InternalDB`. What's crazy to me is that there's already an interface called `cloud.KMSEnv` which we then implement on top of the ExecCfg in this utils package. It's so silly. Also, the `cloud.ExternalStorageContext` has a lot of this stuff. It seems like this is called in exactly one place. There must be a way to unwind this mess. 


<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
*[`pkg/sql/telemetry_logging_test.go` line 11 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NNDBjw0D1uGQ9qNcosm:-NNHyPwc7vBYS3lwztLI:b49u7qt) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d1147f337155975914027ccf1514643242fd6ca/pkg/sql/telemetry_logging_test.go#L11)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This dependency is absolutely wild. We should nix it. The dependency is on the `execCfg` which is used for 3 fields: `Settings`, `ExternalIODir` and `InternalDB`. What's crazy to me is that there's already an interface called `cloud.KMSEnv` which we then implement on top of the ExecCfg in this utils package. It's so silly. Also, the `cloud.ExternalStorageContext` has a lot of this stuff. It seems like this is called in exactly one place. There must be a way to unwind this mess. 
</blockquote></details>

Thanks! I'll push a PR removing dep on `sql` from `externalconn/utils` shortly.


<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
*[`pkg/sql/telemetry_logging_test.go` line 11 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NNDBjw0D1uGQ9qNcosm:-NNbg4kFDyU6LvWD8uhu:bplutvh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d1147f337155975914027ccf1514643242fd6ca/pkg/sql/telemetry_logging_test.go#L11)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

Thanks! I'll push a PR removing dep on `sql` from `externalconn/utils` shortly.
</blockquote></details>

Made a PR for this:  https://github.com/cockroachdb/cockroach/pull/96484


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
lgtm! just requesting some adjustments to comments

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-:-NOAewAQ7ER3QAVQB2H0:bifrw78)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @DrewKimball, @msbutler, and @ZhouXing19)

___
*[`pkg/jobs/resultcols.go` line 21 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NOAeUhL5p9wsPZZOCT9:-NOAeUhL5p9wsPZZOCTA:b-mu4fzd) ([raw file](https://github.com/cockroachdb/cockroach/blob/b74b23c4eed51516d30278ed903a3a93b1cc4893/pkg/jobs/resultcols.go#L21)):*
> ```Go
> // (BACKUP, RESTORE, IMPORT, etc) stmt results.
> // If this header is updated, please also update pgwire.jobIdColIdx and
> // pgwire.rowsColIdx.
> ```

nit: we no longer use rowsColIdx.

also is the comment needed, now that you added the `init` logic?
___
*[`pkg/jobs/resultcols.go` line 34 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NOAeXBtAXhbjgd-2BBz:-NOAeXBtAXhbjgd-2BC-:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/b74b23c4eed51516d30278ed903a3a93b1cc4893/pkg/jobs/resultcols.go#L34)):*
> ```Go
> // job executes in detached mode (i.e. the caller doesn't wait for job completion).
> // If this header is updated, please also update pgwire.jobIdColIdx and
> // pgwire.rowsColIdx.
> ```

ditto
___
*[`pkg/sql/pgwire/command_result.go` line 629 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NOAehvP97EtgHN6aZJG:-NOAehvP97EtgHN6aZJH:b-ys6mth) ([raw file](https://github.com/cockroachdb/cockroach/blob/b74b23c4eed51516d30278ed903a3a93b1cc4893/pkg/sql/pgwire/command_result.go#L629)):*
> ```Go
> 
> func init() {
> 	jobIdIdxInBulkJobExecutionResultHeader := -1
> ```

can you add a comment explaining why we do this in `init`?


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95042)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @DrewKimball, @msbutler, and @rafiss)

___
*[`pkg/sql/pgwire/command_result.go` line 629 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NOAehvP97EtgHN6aZJG:-NOAq4bL89-AcQjEkI69:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b74b23c4eed51516d30278ed903a3a93b1cc4893/pkg/sql/pgwire/command_result.go#L629)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

can you add a comment explaining why we do this in `init`?
</blockquote></details>

Done.
___
*[`pkg/jobs/resultcols.go` line 21 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NOAeUhL5p9wsPZZOCT9:-NOAq6Ym-8vMRrODqyGC:ba3dfob) ([raw file](https://github.com/cockroachdb/cockroach/blob/b74b23c4eed51516d30278ed903a3a93b1cc4893/pkg/jobs/resultcols.go#L21)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: we no longer use rowsColIdx.

also is the comment needed, now that you added the `init` logic?
</blockquote></details>

Removed
___
*[`pkg/jobs/resultcols.go` line 34 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/95042#-NOAeXBtAXhbjgd-2BBz:-NOAq79A7hV57Rqi_Xda:ba3dfob) ([raw file](https://github.com/cockroachdb/cockroach/blob/b74b23c4eed51516d30278ed903a3a93b1cc4893/pkg/jobs/resultcols.go#L34)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Removed


<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
TFTR!
bors r=rafiss
